_drafts/memfault-esp-component.md

```diff
@@ -0,0 +1,219 @@
+      revision: 1.11.5
+```
+
+Add the `CONFIG_MEMFAULT=y` Kconfig setting, and the Memfault SDK is enabled in
+your project.
+
+This is great! and it's the user experience we'd love for all our supported
+platforms (looking at you, Eclipse-based IDEs... 😁).
+
+## How we go from manual mode to component
+
+### Let me tell you about our existing integration
+
+[Insert steve zissou gif]
+
+ESP-IDF's build system is based on CMake, and has a LOT of features and options.
+
+The manual is here if you're interested in going on a deep dive:
+
+- [ESP-IDF CMake Build System](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system.html)
+
+From our perspective, there's essentially 4 steps that happen when you build an
+ESP-IDF based project[^1] (names are my own):
+
+![Picture of 4 steps of esp-idf build system](/img/esp-component/idf.py-build-system.excalidraw.svg)
+
+1. Scan and load all selected components- each component's `CMakeLists.txt` file
+   is processed in script mode
+1. `Kconfig` files are loaded from ESP-IDF and all registerd components.
+   `Kconfig` settings are loaded (from `sdkconfig` or `sdkconfig.defaults` if
+   `sdkconfig` is not already generated), and the final configuration is
+   generated
+1. Process the component `CMakeLists.txt` files by including the component
+   directories. This loads all the component source files, etc.
+1. Build the project, producing the output artifacts (`.elf` files, OTA images,
+   etc).
+
+Memfault's existing integration enables our component by adding the directory to
+the `EXTRA_COMPONENT_DIRS` CMake variable, which causes the SDK folder to be
+processed, and we call `register_component()` in our component `CMakeLists.txt`
+to add the SDK to the build.
+
+Note that we're using `register_component()` here, instead of the modern
+`idf_component_register()` function (see
+[documentation](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/build-system.html#minimal-component-cmakelists)),
+because we need to support older versions of ESP-IDF. More about that later!
+
+Otherwise, our SDK component is pretty standard- we have a `CMakeLists.txt` file
+and `Kconfig` file that define the SDK's build and configuration settings. Take
+a look here if you're curious:
+
+- [CMakeLists.txt](https://github.com/memfault/memfault-firmware-sdk/blob/master/ports/esp_idf/memfault/CMakeLists.txt)
+- [Kconfig](https://github.com/memfault/memfault-firmware-sdk/blob/master/ports/esp_idf/memfault/Kconfig)
+
+### What we need to do to make this a component
+
+ESP Component Registry-enabled components have the follow properties:
+
+- Root-level `idf_component.yml` manifest
+- Root-level `CMakeLists.txt` and `Kconfig` files
+- The enclosing folder name must _exactly_ match the component name
+
+We have essentially 2 options when it comes to making our SDK an ESP-IDF
+component:
+
+1. **Move the SDK into a new folder that matches the component name**:
```

Comment:
This is very inaccurate- it's more like "reorganize our SDK to put the esp port at the root, and the core files under src etc"